lib/AST/Module.cpp

```diff
@@ -4245,6 +4246,50 @@ ASTNode GetSourceFileAsyncNode::evaluate(Evaluator &eval,
   return ASTNode();
 }
 
+ArrayRef<TypeDecl *> SourceFile::getLocalTypeDecls() const {
+  auto *mutableThis = const_cast<SourceFile *>(this);
+  return evaluateOrDefault(getASTContext().evaluator,
+                           LocalTypeDeclsRequest{mutableThis}, {});
+}
+
+namespace {
+class LocalTypeDeclCollector : public ASTWalker {
+  SmallVectorImpl<TypeDecl *> &results;
+
+public:
+  LocalTypeDeclCollector(SmallVectorImpl<TypeDecl *> &results)
+      : results(results) {}
+
+  MacroWalking getMacroWalkingBehavior() const override {
+    return MacroWalking::Expansion;
+  }
+
+  PreWalkAction walkToDeclPre(Decl *D) override {
+    switch (D->getKind()) {
+    case DeclKind::Enum:
+    case DeclKind::Struct:
+    case DeclKind::Class:
+    case DeclKind::Protocol:
+    case DeclKind::TypeAlias:
```

Comment:
I'm mot really sure we need `TypeAlias` here. IRGen and SILGen don't seem to be handle them.

Reply:
It seems reasonable to keep type aliases here so that this request returns all local types, even if IRGen and SILGen don't care. Debugging and module emission probably care, for example.